astro/main.py

```diff
@@ -171,151 +249,144 @@ def _sleep(self, timestamp):
+
+    def _build_execution_plan(self, now, date):
+        try:
+            data = requests.get('http://api.sunrise-sunset.org/json?lat={0}&lng={1}&date={2}&formatted=0'.format(
+                self._latitude, self._longitude, date.strftime('%Y-%m-%d')
+            )).json()
+            if data['status'] != 'OK':
+                raise RuntimeError(data['status'])
+            execution_plan = {}
+            field_map = {'solar noon': 'solar_noon',
+                         'sunset': 'sunset',
+                         'civil dawn': 'civil_twilight_end',
+                         'nautical dawn': 'nautical_twilight_end',
+                         'astronomical dawn': 'astronomical_twilight_end',
+                         'astronomical dusk': 'astronomical_twilight_begin',
+                         'nautical dusk': 'nautical_twilight_begin',
+                         'civil dusk': 'civil_twilight_begin',
+                         'sunrise': 'sunrise'}
+            for sun_location in set(self._group_actions.keys()) | set(self._bits.keys()):
+                group_actions = self._group_actions.get(sun_location, [])
+                bits = self._bits.get(sun_location, [])
+                if not group_actions and not bits:
+                    continue
+                date = self._convert(data['results'].get(field_map.get(sun_location, 'x')))
+                if date is None:
+                    continue
+                for entry in group_actions:
+                    entry_date = date + timedelta(minutes=entry['offset'])
+                    if entry_date < now:
+                        continue
+                    date_plan = execution_plan.setdefault(entry_date, [])
+                    date_plan.append({'task': 'group_action',
+                                      'source': '{0}{1}'.format(
+                                          sun_location,
+                                          Astro._format_offset(entry['offset'])
+                                      ),
+                                      'data': {'group_action_id': entry['group_action_id']}})
+                for entry in bits:
+                    entry_date = date + timedelta(minutes=entry['offset'])
+                    if entry_date < now:
+                        continue
+                    date_plan = execution_plan.setdefault(entry_date, [])
+                    date_plan.append({'task': 'bit',
+                                      'source': '{0}{1}'.format(
+                                          sun_location,
+                                          Astro._format_offset(entry['offset'])
+                                      ),
+                                      'data': {'action': entry['action'],
+                                               'bit_id': entry['bit_id']}})
+            self._execution_plan = execution_plan
+        except Exception as ex:
+            self.logger('Could not load data: {0}'.format(ex))
```

Comment:
in general, I like the concept of a configurable debug/info logger as we now have for most plugins. Perhaps it's a small effort to include this?